sycl/doc/EnvironmentVariables.md

```diff
@@ -52,6 +53,14 @@ The device indices are zero-based and are unique only within a backend. Therefor
 
 Additionally, if a sub-device is chosen (via numeric index or wildcard), then an additional layer of partitioning can be specified. In other words, a sub-sub-device can be selected. Like sub-devices, this is done with a period ( `.` ) and a sub-sub-device specifier which is a wildcard symbol ( `*` ) or a numeric index.  Example `ONEAPI_DEVICE_SELECTOR=level_zero:0.*.*` would partition device 0 into sub-devices and then partition each of those into sub-sub-devices. The range of grandchild sub-sub-devices would be the final devices available to the app, neither device 0, nor its child partitions would be in that list. 
 
+Lastly, a filter in the grammar can be thought of as a term in conjuction with an action that is taken on all devices that are selected by the term. The action can be an accept action or a discard action.
+The string `<term>` represents an accept action and the string `!<term>` represents a discard action. The underlying term is the same but they perform different actions on the matching devices list.
+For example, `!opencl:*` discards all devices of the opencl backend from the list of available devices. When one or more filters accept a device and one or more filters discard the device, the latter
+have priority and the device is ultimately not made available to the user regadless of the order in which the filter appear. This allows the user to provide selector strings such as `*:gpu;!cuda:*` that accepts all gpu devices except those with a CUDA backend.
+Furthermore, if the value of this enviroment variable only has discarding filters, an accepting filter that matches all devices, but not sub-devices and sub-sub-devices, will be implicitly included in the 
+environment to allow the user to specify only the list of devices that must not be made available. Therefore, `!*:cpu` will accept all devices except those that are of the cpu type and `opencl:*;!*:cpu` 
+will accept all devices of the opencl backend exept those that are of the opencl backend and of the cpu type. Finally, it is valid to have no accepting filter match a device but have at least once
+discarding filters match the device. In this case, the device remains unavailable as expected.
```

Comment:
> When one or more filters accept a device and one or more filters discard the device, the latter
have priority and the device is ultimately not made available to the user regadless of the order in which the filter appear.

This part of the semantic seems weird to me.  I think most people will expect the filters to be applied left-to-right, thus a positive filter following a negative one would add devices into the selected set.  (E.g. `!level_zero:gpu;level_zero:1` would make device `1` available even if it was a GPU.)

I think there are two ways to address this.  The first is to say that the filters are applied left-to-right as I say above.

Alternatively, we could say that negative filters must always occur last in the sequence.  If the user specifies a positive filter following a negative one, the runtime would diagnose an error.  This might be better because it makes the cases that are potentially confusing illegal.

We should also loop @RaviNarayanaswamy into this conversation because OpenMP must also agree to whatever semantic we decide upon.

P.S.: You should run spell check on your text.  I see at least a couple spelling mistakes.

Reply:
I see your point and I agree. I also like the second approach more where we enforce that all the negative filters appear at the end of the selector string. It aligns well with how a user would naturally specify the filters in a selector string by saying what to include first and what to exclude later to reduce mistakes. The first approach would require the user to constantly keep an eye on the order in which filters appear otherwise the user may see unintended behavior and so this approach is more error-prone in my opinion.